File: apps/web/src/profileLearning.ts
import type {
  ResearchProfileConcept,
  ResearchProfileProposal,
  ResearchProfileProposalHistoryEvent,
  ResearchProfileProposalTarget,
  ResearchProfileProposalValue,
  ResearchProfileRecord
} from "./companionClient";

export const PROPOSAL_TYPE_LABELS: Record<ResearchProfileProposal["type"], string> = {
  changed_concept_weights: "Changed concept weights",
  new_search_terms: "New search terms",
  exclusions: "Proposed exclusions",
  preferred_methods: "Preferred methods",
  positive_semantic_examples: "Positive semantic examples",
  negative_semantic_examples: "Negative semantic examples",
  revised_screening_instructions: "Revised screening instructions"
};

export const PROPOSAL_TARGET_LABELS: Record<ResearchProfileProposalTarget, string> = {
  concepts: "Concept weights",
  search_queries: "Search queries",
  exclusions: "Exclusions",
  preferred_evidence_types: "Preferred evidence types"
};

export type ProposalAction =
  | { kind: "accept" | "reject" | "reverse"; proposalId: string }
  | { kind: "modify"; proposalId: string; value: ResearchProfileProposalValue };

const TYPE_TO_TARGET: Partial<Record<ResearchProfileProposal["type"], ResearchProfileProposalTarget>> = {
  changed_concept_weights: "concepts",
  new_search_terms: "search_queries",
  exclusions: "exclusions",
  preferred_methods: "preferred_evidence_types"
};

function clone<T>(value: T): T {
  return JSON.parse(JSON.stringify(value)) as T;
}

export function isActionableProposal(proposal: ResearchProfileProposal): boolean {
  return Boolean(
    TYPE_TO_TARGET[proposal.type] &&
      proposal.target_field &&
      proposal.current_value !== undefined &&
      proposal.proposed_value !== undefined
  );
}

export function proposalTarget(proposal: ResearchProfileProposal): ResearchProfileProposalTarget | null {
  return proposal.target_field ?? TYPE_TO_TARGET[proposal.type] ?? null;
}

export function proposalTitle(proposal: ResearchProfileProposal): string {
  return PROPOSAL_TYPE_LABELS[proposal.type] ?? "Proposed profile update";
}

export function profileValue(record: ResearchProfileRecord, target: ResearchProfileProposalTarget): ResearchProfileProposalValue {
  if (target === "concepts") return clone(record.concepts ?? []);
  return { values: clone(record[target] ?? []) };
}

export function valuesEqual(left: unknown, right: unknown): boolean {
  return JSON.stringify(left) === JSON.stringify(right);
}

function validateConcepts(value: ResearchProfileProposalValue, label: string): string | null {
  if (!Array.isArray(value)) return `${label} must be a list of concepts.`;
  const seen = new Set<string>();
  for (const concept of value) {
    if (!concept || typeof concept !== "object" || typeof concept.term !== "string" || !concept.term.trim()) {
      return `${label} contains an empty concept term.`;
    }
    const key = concept.term.trim().toLocaleLowerCase();
    if (seen.has(key)) return `${label} contains duplicate concepts.`;
    seen.add(key);
    if (concept.weight !== undefined && (!Number.isFinite(concept.weight) || typeof concept.weight === "boolean")) {
      return `${label} weights must be finite numbers.`;
    }
  }
  return null;
}

function validateValues(value: ResearchProfileProposalValue, label: string): string | null {
  if (!value || Array.isArray(value) || typeof value !== "object" || !Array.isArray(value.values)) {
    return `${label} must be a list of values.`;
  }
  const seen = new Set<string>();
  for (const item of value.values) {
    if (typeof item !== "string" || !item.trim()) return `${label} contains an empty value.`;
    const key = item.trim().toLocaleLowerCase();
    if (seen.has(key)) return `${label} contains duplicate values.`;
    seen.add(key);
  }
  return null;
}

export function validateProposalValue(
  target: ResearchProfileProposalTarget,
  value: ResearchProfileProposalValue,
  label: string
): string | null {
  return target === "concepts" ? validateConcepts(value, label) : validateValues(value, label);
}

export function validateProposalForApplication(
  record: ResearchProfileRecord,
  proposal: ResearchProfileProposal,
  value: ResearchProfileProposalValue
): string | null {
  if (proposal.status !== "proposed") return "This proposal has already been decided.";
  const target = proposalTarget(proposal);
  if (!target || !isActionableProposal(proposal)) return "This proposal does not contain an actionable durable value.";
  if (target !== TYPE_TO_TARGET[proposal.type]) return "This proposal type is incompatible with its target field.";
  const valueError = validateProposalValue(target, value, "The proposed value");
  if (valueError) return valueError;
  if (!valuesEqual(profileValue(record, target), proposal.current_value)) {
    return "The profile field changed after this proposal was prepared. Reload and review the proposal before applying it.";
  }
  return null;
}

export function applyProposalValue(
  record: ResearchProfileRecord,
  proposal: ResearchProfileProposal,
  value: ResearchProfileProposalValue
): { record: ResearchProfileRecord; appliedValue: ResearchProfileProposalValue } {
  const target = proposalTarget(proposal);
  if (!target) throw new Error("The proposal has no supported target field.");
  const next = clone(record);
  if (target === "concepts") {
    next.concepts = clone(value as ResearchProfileConcept[]);
    return { record: next, appliedValue: clone(next.concepts) };
  }
  const additions = (value as { values: string[] }).values;
  const existing = [...(next[target] ?? [])];
  const seen = new Set(existing.map((item) => item.toLocaleLowerCase()));
  for (const item of additions) {
    if (!seen.has(item.toLocaleLowerCase())) {
      existing.push(item);
      seen.add(item.toLocaleLowerCase());
    }
  }
  next[target] = existing;
  return { record: next, appliedValue: { values: clone(existing) } };
}

function historyEvent(
  event: ResearchProfileProposalHistoryEvent["event"],
  status: ResearchProfileProposal["status"],
  value: ResearchProfileProposalValue | undefined,
  revision: string,
  note?: string
): ResearchProfileProposalHistoryEvent {
  return {
    event,
    status,
    occurred_at: new Date().toISOString(),
    ...(value === undefined ? {} : { value: clone(value) }),
    revision,
    ...(note ? { note } : {})
  };
}

export function buildDecisionRecord(
  record: ResearchProfileRecord,
  proposal: ResearchProfileProposal,
  action: Exclude<ProposalAction, { kind: "reverse" }>,
  expectedRevision: string
): ResearchProfileRecord {
  const now = new Date().toISOString();
  const proposals = (record.proposals ?? []).map((item) => {
    if (item.proposal_id !== proposal.proposal_id) return item;
    if (action.kind === "reject") {
      return {
        ...item,
        status: "rejected" as const,
        decision_at: now,
        history: [...(item.history ?? []), historyEvent("rejected", "rejected", item.proposed_value, expectedRevision)]
      };
    }
    const value = action.kind === "modify" ? action.value : item.proposed_value;
    if (value === undefined) throw new Error("The proposal has no proposed value.");
    const applied = applyProposalValue(record, item, value);
    const status = action.kind === "modify" ? "modified" as const : "accepted" as const;
    return {
      ...item,
      status,
      modified_value: action.kind === "modify" ? clone(value) : item.modified_value,
      applied_value: clone(applied.appliedValue),
      decision_at: now,
      applied_revision: expectedRevision,
      reversal_result: undefined,
      history: [...(item.history ?? []), historyEvent(status, status, applied.appliedValue, expectedRevision)]
    };
  });
  return { ...record, ...((() => {
    const proposalTargetField = proposalTarget(proposal);
    if (!proposalTargetField || action.kind === "reject") return {};
    const value = action.kind === "modify" ? action.value : proposal.proposed_value;
    if (value === undefined) return {};
    return applyProposalValue(record, proposal, value).record;
  })()), proposals, updated_at: now };
}

export function buildReversalRecord(
  record: ResearchProfileRecord,
  proposal: ResearchProfileProposal,
  expectedRevision: string
): { record: ResearchProfileRecord; blocked: boolean } {
  const target = proposalTarget(proposal);
  if (!target || proposal.applied_value === undefined || proposal.current_value === undefined) {
    throw new Error("This proposal cannot be reversed because its applied snapshot is unavailable.");
  }
  const now = new Date().toISOString();
  const current = profileValue(record, target);
  if (!valuesEqual(current, proposal.applied_value)) {
    const blockedProposal: ResearchProfileProposal = {
      ...proposal,
      reversal_result: "blocked",
      history: [
        ...(proposal.history ?? []),
        historyEvent(
          "reversal_blocked",
          proposal.status,
          current,
          expectedRevision,
          "The profile field changed after this proposal was applied."
        )
      ]
    };
    return {
      record: {
        ...record,
        proposals: (record.proposals ?? []).map((item) => item.proposal_id === proposal.proposal_id ? blockedProposal : item),
        updated_at: now
      },
      blocked: true
    };
  }
  const next = clone(record);
  if (target === "concepts") next.concepts = clone(proposal.current_value as ResearchProfileConcept[]);
  else next[target] = clone((proposal.current_value as { values: string[] }).values);
  next.proposals = (record.proposals ?? []).map((item) => item.proposal_id === proposal.proposal_id ? {
    ...item,
    status: "reversed" as const,
    reversed_at: now,
    reversal_result: "restored" as const,
    history: [...(item.history ?? []), historyEvent("reversed", "reversed", proposal.current_value, expectedRevision)]
  } : item);
  next.updated_at = now;
  return { record: next, blocked: false };
}

File: companion/src/research_intelligence_companion/profile_learning.py
from __future__ import annotations

import math
from copy import deepcopy
from typing import Any

SUPPORTED_PROPOSAL_TYPES = {
    "changed_concept_weights",
    "new_search_terms",
    "exclusions",
    "preferred_methods",
}

PROPOSAL_TARGET_FIELDS = {
    "concepts",
    "search_queries",
    "exclusions",
    "preferred_evidence_types",
}

PROPOSAL_TYPE_TO_FIELD = {
    "changed_concept_weights": "concepts",
    "new_search_terms": "search_queries",
    "exclusions": "exclusions",
    # Task 3C maps the proposal's user-facing "preferred methods" language
    # to the existing durable evidence-type field.
    "preferred_methods": "preferred_evidence_types",
}

_PROPOSAL_EVENTS = {
    "created",
    "accepted",
    "modified",
    "rejected",
    "reversed",
    "reversal_blocked",
}
_PROPOSAL_STATUSES = {"proposed", "accepted", "modified", "rejected", "reversed"}


def _is_task3c_proposal(proposal: dict[str, Any]) -> bool:
    return "target_field" in proposal or any(
        key in proposal
        for key in (
            "current_value",
            "proposed_value",
            "modified_value",
            "applied_value",
            "history",
        )
    )


def _validate_string_values(value: Any, label: str) -> None:
    if (
        not isinstance(value, dict)
        or set(value) != {"values"}
        or not isinstance(value["values"], list)
    ):
        raise ValueError(f"{label} must be an object containing a values list.")
    seen: set[str] = set()
    for item in value["values"]:
        if not isinstance(item, str) or not item.strip():
            raise ValueError(f"{label} values must be non-empty strings.")
        normalised = item.strip().casefold()
        if normalised in seen:
            raise ValueError(f"{label} values must not contain case-insensitive duplicates.")
        seen.add(normalised)


def _validate_concept_values(value: Any, label: str) -> None:
    if not isinstance(value, list):
        raise ValueError(f"{label} must be a list of weighted concepts.")
    seen: set[str] = set()
    for item in value:
        if not isinstance(item, dict) or set(item) - {"term", "weight"} or "term" not in item:
            raise ValueError(f"{label} contains an invalid concept.")
        term = item["term"]
        if not isinstance(term, str) or not term.strip():
            raise ValueError(f"{label} terms must be non-empty strings.")
        normalised = term.strip().casefold()
        if normalised in seen:
            raise ValueError(f"{label} terms must not contain case-insensitive duplicates.")
        seen.add(normalised)
        if "weight" in item and (
            isinstance(item["weight"], bool) or not isinstance(item["weight"], (int, float))
        ):
            raise ValueError(f"{label} weights must be finite numbers.")
        if "weight" in item and not math.isfinite(float(item["weight"])):
            raise ValueError(f"{label} weights must be finite numbers.")


def _validate_value(target_field: str, value: Any, label: str) -> None:
    if target_field == "concepts":
        _validate_concept_values(value, label)
    else:
        _validate_string_values(value, label)


def _validate_history(proposal: dict[str, Any]) -> None:
    history = proposal.get("history", [])
    if not isinstance(history, list):
        raise ValueError("Proposal history must be a list.")
    for event in history:
        if not isinstance(event, dict) or event.get("event") not in _PROPOSAL_EVENTS:
            raise ValueError("Proposal history contains an invalid event.")
        if "occurred_at" not in event:
            raise ValueError("Proposal history events require occurred_at.")


def validate_profile_proposals(
    record: dict[str, Any],
    previous_record: dict[str, Any] | None = None,
) -> None:
    proposals = record.get("proposals", [])
    if not isinstance(proposals, list):
        raise ValueError("Research Profile proposals must be a list.")
    identifiers: set[str] = set()
    for proposal in proposals:
        if not isinstance(proposal, dict):
            raise ValueError("Research Profile proposals must be objects.")
        proposal_id = proposal.get("proposal_id")
        if proposal_id in identifiers:
            raise ValueError("Research Profile proposal IDs must be unique.")
        identifiers.add(proposal_id)
        if not _is_task3c_proposal(proposal):
            # Preserve old Task 3B-compatible proposal shells without guessing
            # the missing durable values needed for an actionable decision.
            continue
        proposal_type = proposal.get("type")
        target_field = proposal.get("target_field")
        if proposal_type not in SUPPORTED_PROPOSAL_TYPES:
            raise ValueError("This proposal type is not supported by Task 3C.")
        if (
            target_field not in PROPOSAL_TARGET_FIELDS
            or PROPOSAL_TYPE_TO_FIELD[proposal_type] != target_field
        ):
            raise ValueError("Proposal type and target field do not match.")
        for key in ("current_value", "proposed_value"):
            if key not in proposal:
                raise ValueError(f"Task 3C proposals require {key}.")
            _validate_value(target_field, proposal[key], key)
        if "modified_value" in proposal:
            _validate_value(target_field, proposal["modified_value"], "modified_value")
        if "applied_value" in proposal:
            _validate_value(target_field, proposal["applied_value"], "applied_value")
        _validate_history(proposal)
        status = proposal.get("status")
        if status not in _PROPOSAL_STATUSES:
            raise ValueError("Proposal status is invalid.")
        if status in {"accepted", "modified"}:
            required = {"decision_at", "applied_revision", "applied_value"}
            if not required.issubset(proposal):
                raise ValueError("Applied proposals require decision, revision, and applied value.")
        if status == "modified" and "modified_value" not in proposal:
            raise ValueError("Modified proposals require modified_value.")
        if status == "rejected" and "decision_at" not in proposal:
            raise ValueError("Rejected proposals require decision_at.")
        if status == "reversed":
            required = {"decision_at", "applied_revision", "applied_value", "reversed_at"}
            if not required.issubset(proposal) or proposal.get("reversal_result") != "restored":
                raise ValueError("Reversed proposals require a successful reversal record.")
        if proposal.get("reversal_result") == "blocked" and status not in {"accepted", "modified"}:
            raise ValueError("A blocked reversal must preserve an accepted or modified status.")

    if previous_record is not None:
        _validate_proposal_transition(previous_record.get("proposals", []), proposals)


def _validate_proposal_transition(previous: Any, incoming: Any) -> None:
    if not isinstance(previous, list) or not isinstance(incoming, list):
        return
    incoming_by_id = {item.get("proposal_id"): item for item in incoming if isinstance(item, dict)}
    for old in previous:
        if not isinstance(old, dict):
            continue
        proposal_id = old.get("proposal_id")
        new = incoming_by_id.get(proposal_id)
        if new is None:
            raise ValueError("Proposal decision history cannot be deleted.")
        for key in (
            "proposal_id",
            "type",
            "explanation",
            "created_at",
            "target_field",
            "current_value",
            "proposed_value",
        ):
            if old.get(key) != new.get(key):
                raise ValueError("Proposal identity and original values cannot be changed.")
        old_status = old.get("status")
        new_status = new.get("status")
        if old_status == "proposed" and new_status not in {
            "proposed",
            "accepted",
            "modified",
            "rejected",
        }:
            raise ValueError(
                "A proposed profile change can only be accepted, modified, or rejected."
            )
        if old_status in {"accepted", "modified"} and new_status not in {old_status, "reversed"}:
            raise ValueError("An applied proposal cannot be decided a second time.")
        if old_status in {"accepted", "modified"} and new_status == old_status:
            for key in (
                "decision_at",
                "modified_value",
                "applied_value",
                "applied_revision",
                "reversed_at",
            ):
                if old.get(key) != new.get(key):
                    raise ValueError("An applied proposal decision is immutable.")
            if old.get("reversal_result") != new.get("reversal_result") and new.get(
                "reversal_result"
            ) != "blocked":
                raise ValueError("An applied proposal decision is immutable.")
        if old_status in {"rejected", "reversed"} and new_status != old_status:
            raise ValueError("A completed proposal decision cannot be changed.")


def profile_value(record: dict[str, Any], target_field: str) -> Any:
    if target_field == "concepts":
        return deepcopy(record.get("concepts", []))
    return {"values": deepcopy(record.get(target_field, []))}


def validate_profile_value(target_field: str, value: Any) -> None:
    _validate_value(target_field, value, "profile value")
